pytissueoptics/scene/tests/intersection/testBoundingBoxIntersect.py

```diff
@@ -0,0 +1,119 @@
+import unittest
+
+from ddt import ddt, data
+
+from pytissueoptics.scene.geometry import Vector, BoundingBox
+from pytissueoptics.scene.intersection import Ray
+from pytissueoptics.scene.intersection.bboxIntersect import GemsBoxIntersect, ZacharBoxIntersect
+
+
+@ddt
+class TestBoxIntersect(unittest.TestCase):
+    intersectStrategies = [GemsBoxIntersect, ZacharBoxIntersect]
+
+    @data(*intersectStrategies)
+    def testGivenIntersectingRayAndBox_shouldReturnClosestIntersectionPoint(self, IntersectStrategy):
+        box = BoundingBox([0 + 2, 1 + 2], [0, 1], [-1, 0])
+        rayOrigin = Vector(0.25 + 2, 0.25, 2)
+        rayDirection = Vector(0.1, 0, -1)
+        rayDirection.normalize()
+        ray = Ray(rayOrigin, rayDirection)
+
+        intersection = IntersectStrategy().getIntersection(ray, box)
+
+        self.assertIsNotNone(intersection)
+        self.assertEqual(2.45, intersection.x)
+        self.assertEqual(0.25, intersection.y)
+        self.assertEqual(0.0, intersection.z)
+
+    @data(*intersectStrategies)
+    def testGivenNonIntersectingRayAndBox_shouldReturnNone(self, IntersectStrategy):
+        box = BoundingBox([0, 1], [0, 1], [-1, 0])
+        rayOrigin = Vector(0.25, 0.25, 2)
+        rayDirection = Vector(-0.2, 0, -1)
+        rayDirection.normalize()
+        ray = Ray(rayOrigin, rayDirection)
+
+        intersection = IntersectStrategy().getIntersection(ray, box)
+
+        self.assertIsNone(intersection)
+
+    @data(*intersectStrategies)
+    def testGivenRayPointingAwayFromTheBox_shouldReturnNone(self, IntersectStrategy):
+        box = BoundingBox([0, 1], [0, 1], [-1, 0])
+        rayOrigin = Vector(0.25, 0.25, 2)
+        rayDirection = Vector(-0.1, 0, 1)
+        rayDirection.normalize()
+        ray = Ray(rayOrigin, rayDirection)
+
+        intersection = IntersectStrategy().getIntersection(ray, box)
+
+        self.assertIsNone(intersection)
+
+    @data(*intersectStrategies)
+    def testGivenRayInsideBox_shouldReturnRayOrigin(self, IntersectStrategy):
+        box = BoundingBox([0, 1], [0, 1], [-2, 1])
+        rayOrigin = Vector(0.25, 0.25, 0)
+        rayDirection = Vector(0.1, 0, -1)
+        rayDirection.normalize()
+        ray = Ray(rayOrigin, rayDirection)
+
+        intersection = IntersectStrategy().getIntersection(ray, box)
+        
+        self.assertEqual(ray.origin, intersection)
```

Comment:
Not exactly what I would expect from an intersector, but the clients do need a special return value if the ray starts inside. I could also raise a custom error to make sure the client is aware of this behavior, but that would require a try-catch in the client's code. Similarly I could add a return type but that would require weird type checks. There might be another solution... For now returning ray.origin is easy to handle but hard to spot (not an expected behavior unless you read the tests).

Reply:
ok